### ISSUE TITLE
Bulk syncing failing due to fork or incorrectly sent data

### ISSUE THREAD

**Issue by mrwillis** [0xPolygon/polygon-edge#180]
### Description

When syncing a node from a fresh state, it seems that it does not know how to handle forks correctly. Block `929697` is when the fork starts. The block hash for `929697` on the canonical chain is `0x107a2916feb361e77ab753db582c8d206c34bc7429370b64396b3eebd48842ac` but `929698` is expecting `0x1492502b1455bb8f53def60027b69ff0b82e4538d7d157d0170952ac1470a95d` which doesn't exist

Is it possible the node it is syncing from is storing an incorrect chain?

### Your environment

* OS and version Debian 11
* version of the Polygon SDK 335e78199b52c768e9b00c3f955e9a42a3ce5ed4
* branch that causes this issue `develop`

### Steps to reproduce

* Tell us how to reproduce this issue <br />
* Where the issue is, if you know <br /> `WriteBlocks()` in `BulkSyncWithPeer()`
* Which commands triggered the issue, if any 

```
/home/$USER/validator/node/main server --data-dir /home/$USER/validator/node/mynode --chain /home/$USER/validator/node/genesis.json --grpc 0.0.0.0:10000 --libp2p 0.0.0.0:10001 --jsonrpc 0.0.0.0:10002 --nat $IP --seal
```

### Expected behaviour

The node should sync normally up to the current block

### Actual behaviour

The node gets stuck on block `929697`

### Logs

Node 1 trying to sync

```
Oct 13 17:52:18 instance-1 main[6539]: 2021-10-13T17:52:18.925Z [INFO]  polygon.blockchain: new block: number=929682 hash=0x273ee8e6b14a410796ca579efd41fcef8730a2ab4755dc02f1fe43c4c67ab67b txns=1 generation_time_in_sec=2
Oct 13 17:52:18 instance-1 main[6539]: 2021-10-13T17:52:18.926Z [INFO]  polygon.blockchain: new block: number=929683 hash=0x0a84b7292f2c578481a1ba60b4f3573f44bf069bedaf4fc6b4f45659a72eb435 txns=0 generation_time_in_sec=10
Oct 13 17:52:18 instance-1 main[6539]: 2021-10-13T17:52:18.926Z [INFO]  polygon.blockchain: new block: number=929684 hash=0xcacaada2216dc16a2cba2684725e6bd65f73f728e360196e3cd46e79d743959b txns=0 generation_time_in_sec=2
Oct 13 17:52:18 instance-1 main[6539]: 2021-10-13T17:52:18.927Z [INFO]  polygon.blockchain: new block: number=929685 hash=0x5aea6156f8f1c50ae2cee8807b77cdf35c3d84370eaf3ec1c6f80bcf425b4823 txns=0 generation_time_in_sec=2
Oct 13 17:52:18 instance-1 main[6539]: 2021-10-13T17:52:18.928Z [INFO]  polygon.blockchain: new block: number=929686 hash=0x744e81c4227c8eb185857701fa98d7e99234085111daae5162c1835ed87c4dcd txns=0 generation_time_in_sec=2
Oct 13 17:52:18 instance-1 main[6539]: 2021-10-13T17:52:18.928Z [INFO]  polygon.blockchain: new block: number=929687 hash=0x6f7863e883ccc53d69ba1b590936146308188351bbacde35523b3787e3712b90 txns=0 generation_time_in_sec=10
Oct 13 17:52:18 instance-1 main[6539]: 2021-10-13T17:52:18.928Z [INFO]  polygon.blockchain: new head: hash=0x6f7863e883ccc53d69ba1b590936146308188351bbacde35523b3787e3712b90 number=929687
Oct 13 17:52:18 instance-1 main[6539]: 2021-10-13T17:52:18.928Z [INFO]  polygon.blockchain: write blocks: num=10 from=929688 to=929697 parent=0x6f7863e883ccc53d69ba1b590936146308188351bbacde35523b3787e3712b90
Oct 13 17:52:18 instance-1 main[6539]: 2021-10-13T17:52:18.946Z [INFO]  polygon.blockchain: new block: number=929688 hash=0x6e10907711cce5a5bbf958ffe09e3d80a9b31324649af46d2999f1d218ea425d txns=0 generation_time_in_sec=2
Oct 13 17:52:18 instance-1 main[6539]: 2021-10-13T17:52:18.946Z [INFO]  polygon.blockchain: new block: number=929689 hash=0x4005052dee7146a1a0e36c615935cc3d0ae9db7dcb20d5edf4949803431fe3a4 txns=0 generation_time_in_sec=2
Oct 13 17:52:18 instance-1 main[6539]: 2021-10-13T17:52:18.947Z [INFO]  polygon.blockchain: new block: number=929690 hash=0xa92ac10b4fd7e9c20cfcf4e9d4a5b1eb3579f424c153d209ae7d5891082a5971 txns=1 generation_time_in_sec=2
Oct 13 17:52:18 instance-1 main[6539]: 2021-10-13T17:52:18.948Z [INFO]  polygon.blockchain: new block: number=929691 hash=0xdb20e30540903461c0531c364d374668b6c76d2a476cd70d0159f190fbcb0ff3 txns=0 generation_time_in_sec=10
Oct 13 17:52:18 instance-1 main[6539]: 2021-10-13T17:52:18.948Z [INFO]  polygon.blockchain: new block: number=929692 hash=0x808e93329dafa8ef47724e999c99d1e7934328b5b60f4e0b10f2e155c62365ec txns=0 generation_time_in_sec=2
Oct 13 17:52:18 instance-1 main[6539]: 2021-10-13T17:52:18.949Z [INFO]  polygon.blockchain: new block: number=929693 hash=0x1bfcbe5debfd343f358d0e98fa61734fee8b135b14a40771c88df35049fc01a3 txns=0 generation_time_in_sec=2
Oct 13 17:52:18 instance-1 main[6539]: 2021-10-13T17:52:18.949Z [INFO]  polygon.blockchain: new block: number=929694 hash=0xc18950225162eedb6cd6389096f78f1f7ab00be8c0d69ef1eb54f041c8b8e448 txns=0 generation_time_in_sec=2
Oct 13 17:52:18 instance-1 main[6539]: 2021-10-13T17:52:18.950Z [INFO]  polygon.blockchain: new block: number=929695 hash=0xdd0a374030ea24430e32b7a3d1b826db2aac78728c1e6078c0ea3ae6a3b638e8 txns=0 generation_time_in_sec=10
Oct 13 17:52:18 instance-1 main[6539]: 2021-10-13T17:52:18.950Z [INFO]  polygon.blockchain: new block: number=929696 hash=0xd943d00fffa9d2c84c336db0b1362f624234c8d4716ff587d0c0f69cc43e8f32 txns=0 generation_time_in_sec=2
Oct 13 17:52:18 instance-1 main[6539]: 2021-10-13T17:52:18.951Z [INFO]  polygon.blockchain: new block: number=929697 hash=0x107a2916feb361e77ab753db582c8d206c34bc7429370b64396b3eebd48842ac txns=0 generation_time_in_sec=2
Oct 13 17:52:18 instance-1 main[6539]: 2021-10-13T17:52:18.951Z [INFO]  polygon.blockchain: new head: hash=0x107a2916feb361e77ab753db582c8d206c34bc7429370b64396b3eebd48842ac number=929697
Oct 13 17:52:18 instance-1 main[6539]: 2021-10-13T17:52:18.951Z [INFO]  polygon.blockchain: write blocks: num=10 from=929698 to=929707 parent=0x1492502b1455bb8f53def60027b69ff0b82e4538d7d157d0170952ac1470a95d
Oct 13 17:52:18 instance-1 main[6539]: 2021-10-13T17:52:18.951Z [ERROR] polygon.consensus.ibft: failed to bulk sync: err="failed to write bulk sync blocks: parent of 0xa58235450c120fbad063b0b34a9c0c866bc1289d332390d135a39194c2861e22 (929698) not found: 0x1492502b1455bb8f53def60027b69ff0b82e4538d7d157d0170952ac1470a95d"
Oct 13 17:52:19 instance-1 main[6539]: 2021-10-13T17:52:19.867Z [INFO]  polygon.blockchain: write blocks: num=10 from=929698 to=929707 parent=0x1492502b1455bb8f53def60027b69ff0b82e4538d7d157d0170952ac1470a95d
Oct 13 17:52:19 instance-1 main[6539]: 2021-10-13T17:52:19.867Z [ERROR] polygon.consensus.ibft: failed to bulk sync: err="failed to write bulk sync blocks: parent of 0xa58235450c120fbad063b0b34a9c0c866bc1289d332390d135a39194c2861e22 (929698) not found: 0x1492502b1455bb8f53def60027b69ff0b82e4538d7d157d0170952ac1470a95d"
Oct 13 17:52:20 instance-1 main[6539]: 2021-10-13T17:52:20.774Z [INFO]  polygon.blockchain: write blocks: num=10 from=929698 to=929707 parent=0x1492502b1455bb8f53def60027b69ff0b82e4538d7d157d0170952ac1470a95d
Oct 13 17:52:20 instance-1 main[6539]: 2021-10-13T17:52:20.774Z [ERROR] polygon.consensus.ibft: failed to bulk sync: err="failed to write bulk sync blocks: parent of 0xa58235450c120fbad063b0b34a9c0c866bc1289d332390d135a39194c2861e22 (929698) not found: 0x1492502b1455bb8f53def60027b69ff0b82e4538d7d157d0170952ac1470a95d"
Oct 13 17:52:21 instance-1 main[6539]: 2021-10-13T17:52:21.680Z [INFO]  polygon.blockchain: write blocks: num=10 from=929698 to=929707 parent=0x1492502b1455bb8f53def60027b69ff0b82e4538d7d157d0170952ac1470a95d
Oct 13 17:52:21 instance-1 main[6539]: 2021-10-13T17:52:21.680Z [ERROR] polygon.consensus.ibft: failed to bulk sync: err="failed to write bulk sync blocks: parent of 0xa58235450c120fbad063b0b34a9c0c866bc1289d332390d135a39194c2861e22 (929698) not found: 0x1492502b1455bb8f53def60027b69ff0b82e4538d7d157d0170952ac1470a95d"
Oct 13 17:52:22 instance-1 main[6539]: 2021-10-13T17:52:22.585Z [INFO]  polygon.blockchain: write blocks: num=10 from=929698 to=929707 parent=0x1492502b1455bb8f53def60027b69ff0b82e4538d7d157d0170952ac1470a95d
Oct 13 17:52:22 instance-1 main[6539]: 2021-10-13T17:52:22.586Z [ERROR] polygon.consensus.ibft: failed to bulk sync: err="failed to write bulk sync blocks: parent of 0xa58235450c120fbad063b0b34a9c0c866bc1289d332390d135a39194c2861e22 (929698) not found: 0x1492502b1455bb8f53def60027b69ff0b82e4538d7d157d0170952ac1470a95d"
Oct 13 17:52:23 instance-1 main[6539]: 2021-10-13T17:52:23.493Z [INFO]  polygon.blockchain: write blocks: num=10 from=929698 to=929707 parent=0x1492502b1455bb8f53def60027b69ff0b82e4538d7d157d0170952ac1470a95d
Oct 13 17:52:23 instance-1 main[6539]: 2021-10-13T17:52:23.493Z [ERROR] polygon.consensus.ibft: failed to bulk sync: err="failed to write bulk sync blocks: parent of 0xa58235450c120fbad063b0b34a9c0c866bc1289d332390d135a39194c2861e22 (929698) not found: 0x1492502b1455bb8f53def60027b69ff0b82e4538d7d157d0170952ac1470a95d"
Oct 13 17:52:24 instance-1 main[6539]: 2021-10-13T17:52:24.407Z [INFO]  polygon.blockchain: write blocks: num=10 from=929698 to=929707 parent=0x1492502b1455bb8f53def60027b69ff0b82e4538d7d157d0170952ac1470a95d
Oct 13 17:52:24 instance-1 main[6539]: 2021-10-13T17:52:24.408Z [ERROR] polygon.consensus.ibft: failed to bulk sync: err="failed to write bulk sync blocks: parent of 0xa58235450c120fbad063b0b34a9c0c866bc1289d332390d135a39194c2861e22 (929698) not found: 0x1492502b1455bb8f53def60027b69ff0b82e4538d7d157d0170952ac1470a95d"
Oct 13 17:52:25 instance-1 main[6539]: 2021-10-13T17:52:25.317Z [INFO]  polygon.blockchain: write blocks: num=10 from=929698 to=929707 parent=0x1492502b1455bb8f53def60027b69ff0b82e4538d7d157d0170952ac1470a95d
Oct 13 17:52:25 instance-1 main[6539]: 2021-10-13T17:52:25.317Z [ERROR] polygon.consensus.ibft: failed to bulk sync: err="failed to write bulk sync blocks: parent of 0xa58235450c120fbad063b0b34a9c0c866bc1289d332390d135a39194c2861e22 (929698) not found: 0x1492502b1455bb8f53def60027b69ff0b82e4538d7d157d0170952ac1470a95d"
Oct 13 17:52:26 instance-1 main[6539]: 2021-10-13T17:52:26.224Z [INFO]  polygon.blockchain: write blocks: num=10 from=929698 to=929707 parent=0x1492502b1455bb8f53def60027b69ff0b82e4538d7d157d0170952ac1470a95d
Oct 13 17:52:26 instance-1 main[6539]: 2021-10-13T17:52:26.224Z [ERROR] polygon.consensus.ibft: failed to bulk sync: err="failed to write bulk sync blocks: parent of 0xa58235450c120fbad063b0b34a9c0c866bc1289d332390d135a39194c2861e22 (929698) not found: 0x1492502b1455bb8f53def60027b69ff0b82e4538d7d157d0170952ac1470a95d"
Oct 13 17:52:27 instance-1 main[6539]: 2021-10-13T17:52:27.134Z [INFO]  polygon.blockchain: write blocks: num=10 from=929698 to=929707 parent=0x1492502b1455bb8f53def60027b69ff0b82e4538d7d157d0170952ac1470a95d
Oct 13 17:52:27 instance-1 main[6539]: 2021-10-13T17:52:27.134Z [ERROR] polygon.consensus.ibft: failed to bulk sync: err="failed to write bulk sync blocks: parent of 0xa58235450c120fbad063b0b34a9c0c866bc1289d332390d135a39194c2861e22 (929698) not found: 0x1492502b1455bb8f53def60027b69ff0b82e4538d7d157d0170952ac1470a95d"
```

Node 2 trying to sync (detects a fork)

```
2
2021-10-12T11:12:39.495-07:00
{"message":"2021-10-12T18:12:39.254Z [INFO] polygon.blockchain: write blocks: num=10 from=929698 to=929707 parent=0x1492502b1455bb8f53def60027b69ff0b82e4538d7d157d0170952ac1470a95d"}
3
2021-10-12T11:12:39.495-07:00
{"message":"2021-10-12T18:12:39.255Z [ERROR] polygon.consensus.ibft: failed to bulk sync: err=\"failed to write bulk sync blocks: parent of 0xa58235450c120fbad063b0b34a9c0c866bc1289d332390d135a39194c2861e22 (929698) not found: 0x1492502b1455bb8f53def60027b69ff0b82e4538d7d157d0170952ac1470a95d\""}
4
2021-10-12T11:12:39.495-07:00
{"message":"2021-10-12T18:12:39.273Z [DEBUG] polygon.consensus.syncer: fork found: ancestor=929697"}
5
2021-10-12T11:12:39.495-07:00
{"message":"2021-10-12T18:12:39.273Z [DEBUG] polygon.consensus.syncer: sync up to block: from=929698 to=1093038"}
6
2021-10-12T11:12:39.495-07:00
{"message":"2021-10-12T18:12:39.283Z [INFO] polygon.blockchain: write blocks: num=10 from=929698 to=929707 parent=0x1492502b1455bb8f53def60027b69ff0b82e4538d7d157d0170952ac1470a95d"}
7
2021-10-12T11:12:39.495-07:00
{"message":"2021-10-12T18:12:39.284Z [ERROR] polygon.consensus.ibft: failed to bulk sync: err=\"failed to write bulk sync blocks: parent of 0xa58235450c120fbad063b0b34a9c0c866bc1289d332390d135a39194c2861e22 (929698) not found: 0x1492502b1455bb8f53def60027b69ff0b82e4538d7d157d0170952ac1470a95d\""}
8
2021-10-12T11:12:39.495-07:00
{"message":"2021-10-12T18:12:39.305Z [DEBUG] polygon.consensus.syncer: fork found: ancestor=929697"}
9
2021-10-12T11:12:39.495-07:00
{"message":"2021-10-12T18:12:39.305Z [DEBUG] polygon.consensus.syncer: sync up to block: from=929698 to=1093038"}
10
2021-10-12T11:12:39.495-07:00
{"message":"2021-10-12T18:12:39.316Z [INFO] polygon.blockchain: write blocks: num=10 from=929698 to=929707 parent=0x1492502b1455bb8f53def60027b69ff0b82e4538d7d157d0170952ac1470a95d"}
11
2021-10-12T11:12:39.495-07:00
{"message":"2021-10-12T18:12:39.316Z [ERROR] polygon.consensus.ibft: failed to bulk sync: err=\"failed to write bulk sync blocks: parent of 0xa58235450c120fbad063b0b34a9c0c866bc1289d332390d135a39194c2861e22 (929698) not found: 0x1492502b1455bb8f53def60027b69ff0b82e4538d7d157d0170952ac1470a95d\""}
12
2021-10-12T11:12:39.495-07:00
{"message":"2021-10-12T18:12:39.333Z [DEBUG] polygon.consensus.syncer: fork found: ancestor=929697"}
13
2021-10-12T11:12:39.495-07:00
{"message":"2021-10-12T18:12:39.333Z [DEBUG] polygon.consensus.syncer: sync up to block: from=929698 to=1093038"}
14
2021-10-12T11:12:39.495-07:00
{"message":"2021-10-12T18:12:39.350Z [INFO] polygon.blockchain: write blocks: num=10 from=929698 to=929707 parent=0x1492502b1455bb8f53def60027b69ff0b82e4538d7d157d0170952ac1470a95d"}
15
2021-10-12T11:12:39.495-07:00
{"message":"2021-10-12T18:12:39.350Z [ERROR] polygon.consensus.ibft: failed to bulk sync: err=\"failed to write bulk sync blocks: parent of 0xa58235450c120fbad063b0b34a9c0c866bc1289d332390d135a39194c2861e22 (929698) not found: 0x1492502b1455bb8f53def60027b69ff0b82e4538d7d157d0170952ac1470a95d\""}
16
2021-10-12T11:12:39.495-07:00
{"message":"2021-10-12T18:12:39.368Z [DEBUG] polygon.consensus.syncer: fork found: ancestor=929697"}
17
2021-10-12T11:12:39.495-07:00
{"message":"2021-10-12T18:12:39.368Z [DEBUG] polygon.consensus.syncer: sync up to block: from=929698 to=1093038"}
18
2021-10-12T11:12:39.495-07:00
{"message":"2021-10-12T18:12:39.379Z [INFO] polygon.blockchain: write blocks: num=10 from=929698 to=929707 parent=0x1492502b1455bb8f53def60027b69ff0b82e4538d7d157d0170952ac1470a95d"}
19
2021-10-12T11:12:39.495-07:00
{"message":"2021-10-12T18:12:39.379Z [ERROR] polygon.consensus.ibft: failed to bulk sync: err=\"failed to write bulk sync blocks: parent of 0xa58235450c120fbad063b0b34a9c0c866bc1289d332390d135a39194c2861e22 (929698) not found: 0x1492502b1455bb8f53def60027b69ff0b82e4538d7d157d0170952ac1470a95d\""}
20
2021-10-12T11:12:39.495-07:00
{"message":"2021-10-12T18:12:39.398Z [DEBUG] polygon.consensus.syncer: fork found: ancestor=929697"}
```

**Comment by Kourin1996** [comment]
@mrwillis Hello, Thank you for raising this issue. Sorry for the late update. I'm currently checking this issue, but this issue comes from wrong stored data as you said, I guess. I will get back to you soon.

**Comment by Kourin1996** [comment]
@mrwillis May I know if (1) you setup and start IBFT cluster on current develop (335e781) branch and you haven't pulled after this (2) it's possible to share data dir?

**Comment by mrwillis** [comment]
Hey @Kourin1996 yes branch is 335e78199b52c768e9b00c3f955e9a42a3ce5ed4 I can confirm that this issue happened. Sure I can share. Can you DM me in telegram? You can get my handle from lazar or milos.

**Comment by Kourin1996** [comment]
@mrwillis Hi, Thank you for a message. I've received data from Milos. I appreciate it.

**Comment by dankostiuk** [comment]
Hi @Kourin1996 , after pulling in the latest develop which includes #187 , I am still observing similar log output when syncing a validator node from a fresh state:
```
Nov 04 19:07:45 ip-172-31-32-129 main[4126412]: 2021-11-04T19:07:45.457Z [DEBUG] polygon.consensus.syncer: fork found: ancestor=929697
Nov 04 19:07:45 ip-172-31-32-129 main[4126412]: 2021-11-04T19:07:45.458Z [DEBUG] polygon.consensus.syncer: sync up to block: from=929698 to=1209712
Nov 04 19:07:45 ip-172-31-32-129 main[4126412]: 2021-11-04T19:07:45.468Z [INFO]  polygon.blockchain: write blocks: num=10 from=929698 to=929707 parent=0x1492502b1455bb8f53def60027b69ff0b82e4538d7d157d0170952ac1470a95d
Nov 04 19:07:45 ip-172-31-32-129 main[4126412]: 2021-11-04T19:07:45.469Z [ERROR] polygon.consensus.ibft: failed to bulk sync: err="failed to write bulk sync blocks: parent of 0xa58235450c120fbad063b0b34a9c0c866bc1289d332390d135a39194c2861e22 (929698) not found: 0x1492502b1455bb8f53def60027b69ff0b82e4538d7d157d0170952ac1470a95d"
```

I'm working off of `578673c` (yesterday's develop) running a testnet with 4 validators. I'm happy to share the data dir if you'd like.

**Comment by Kourin1996** [comment]
@dankostiuk Hi, Thank you for reporting. Did you start with new chain? or Did you let new node sync with existing blockchain? Thanks

**Comment by Kourin1996** [comment]
@dankostiuk @mrwillis I apologize for late reply. May I ask several questions?

(1) How many validators and non-validators did you run?
(2) How often did you reset node data when the issue happened? (reset means deleting any data in node and let node sync from genesis)
(3) Did you reset multiple nodes at the same time?
(4) Did all validator nodes connect to each other?

I think blockchain stored in storage is wrong. For that reason, you might want to sync with different peer. I'm investigating why generate wrong chain.

**Comment by Kourin1996** [comment]
@mrwillis Could I get the data-dir of all nodes? I have been taking effort to figure out the issue, but I've not got any reason so far. The questions I'm asking above would be helpful for resolving. But it's impossible to sync even though I fix this issue because the chain in data-dir has fork. But if there is a chain without fork in some nodes. We can recover data correctly.

**Comment by Kourin1996** [comment]
Let me summarize what the issue is for the future.
I received data directory from @mrwillis and check the blockchain there, and I found a fork between `929419` and `929697`. And block `929698` (the next block of last block of the fork) has wrong ParentHash. And also I found stopping block production from `952966` to `952967`.

| Block Height | Timestamp | Comment |
| :---         |     :---     |          :--- |
| 929418  | 1633276693     |     |
|...     |       | 2 days and 5 hours diff |
| 929419     | 1633467434       | Fork began |
|...     |       | Fork continues for 278 blocks |
| 929696    | 1633468545 | |
| 929697 | 1633468547 | Fork ended |
| 929698 | 1633277868 | Mismatch of parent hash |
|...     |       |  |
| 952966 | 1633376323 | |
| ... | | 1 day 1 hour diff |
| 952967 | 1633468698 | |

I guess fork happened due to overwriting of part of chain (`929419` and `929697` in this case) with following steps.
1. Generate blocks from genesis to 952966 
2. Generate folks from 929419 to 929697 (overwrite part of chain)
3. Generate blocks from 952967

Basically IBFT won't generate fork basically. I thought syncing with non-proper node happened. (But I'm not sure this assumption is correct)

**Comment by dankostiuk** [comment]
> @dankostiuk @mrwillis I apologize for late reply. May I ask several questions?
> (1) How many validators and non-validators did you run? 
(2) How often did you reset node data when the issue happened? (reset means deleting any data in node and let node sync from genesis) 
(3) Did you reset multiple nodes at the same time? 
(4) Did all validator nodes connect to each other?
> 
> I think blockchain stored in storage is wrong. For that reason, you might want to sync with different peer. I'm investigating why generate wrong chain.

Hey @Kourin1996, I let the new node sync with the existing chain that had the same conflicting fork block.

(1) - We had 4 validators and 2 non-validators connected - although the `ibft snapshot` command seemed to indicate that there were 6 validators total.

(2) - We always see this issue when doing a complete reset of node data, however when syncing from an S3 backup at block `1209712` (which is past the fork block) we had no issues.

(3) - No, we only reset a single validator (validator-4)

(4) - Yes all peer connections were established.

 >@mrwillis Could I get the data-dir of all nodes? I have been taking effort to figure out the issue, but I've not got any reason so far. The questions I'm asking above would be helpful for resolving. But it's impossible to sync even though I fix this issue because the chain in data-dir has fork. But if there is a chain without fork in some nodes. We can recover data correctly.
 
I've uploaded the data-dirs for all 4 validators to S3, along with the backup state (block 1209712 taken from our 5th node which avoids the issue) which I'll share in our telegram chat with @zivkovicmilos . Thanks @Kourin1996!

**Comment by Kourin1996** [comment]
@dankostiuk Hello, Thank you for sharing data-dir.

> (2) - We always see this issue when doing a complete reset of node data, however when syncing from an S3 backup at block 1209712 (which is past the fork block) we had no issues.

Is it possible to recover from this S3 backup? Just update, I checked all data-dirs but there was a same fork in them.

**Comment by dankostiuk** [comment]
Hey @Kourin1996, yes we've been able to recover fine using the S3 backup. As there have been a lot of changes recently, we've decided to pull the latest and reset our testnet to block 0. We're hoping some of the recent changes fix the issue we've experienced here and we'll be sure to let you know if we see anything similar again.

Thanks for your continued help and feel free to close this issue.
